### ISSUE TITLE
3.14 deferred evaluation of annotations

### ISSUE THREAD

**Issue by nickdrozd** [astral-sh/ruff#15100]
In 3.14, type annotations are not evaluated eagerly.

https://docs.python.org/3.14/whatsnew/3.14.html#whatsnew314-pep649

So this snippet runs fine:

```python
class X:
    x: X
```

Ruff flags the field annotation as an undefined name. It is undefined in 3.13 and earlier, but in 3.14 this is a false positive. Earlier versions require `from __future__ import annotations` to make this work.

**Comment by dhruvmanila** [comment]
Thanks for the issue.

The following are the affected rules that should be updated (not a complete list):
* https://docs.astral.sh/ruff/rules/#flake8-future-annotations-fa
* https://docs.astral.sh/ruff/rules/unnecessary-future-import/

As per the [release schedule of 3.14](https://peps.python.org/pep-0745/), the first beta is scheduled in around 5 months so I'd wait for a while until making any changes for 3.14.

**Comment by Daverball** [comment]
Most importantly we will need to update the semantic model. Everywhere that currently checks for the presence of `from __future__ import annotations` to defer visiting a type expression, will need to also check for a Python version of >= 3.14.

That being said, as far as I know, there's not a 1:1 equivalence in semantics between the deferral strategy for PEP 649 and PEP 563, so we may need to do some things subtly different, to properly model the runtime semantics. Since there are some things that are safe with PEP 563, that aren't safe with PEP 649. Which is also the reason why `from __future__ import annotations` doesn't go away in 3.14 and still has the same semantics as it did before (i.e. you will get string forward references in `__annotations__`).

**Comment by nickdrozd** [comment]
Some discussion here: https://github.com/nedbat/coveragepy/issues/1908

**Comment by nickdrozd** [comment]
Besides undefined name, this also raises false positives for TC004.

**Comment by Tatsh** [comment]
I would love to have a rule and fix that says to add `from __future__ import annotations` to the file (and it would be great if the checker checks that it actually makes a difference), and then an upgrade rule/fix when we're on 3.14+ to remove it.

**Comment by Daverball** [comment]
> I would love to have a rule and fix that says to add `from __future__ import annotations` to the file (and it would be great if the checker checks that it actually makes a difference), and then an upgrade rule/fix when we're on 3.14+ to remove it.

You can already force an import with [missing-required-import](https://docs.astral.sh/ruff/rules/missing-required-import/):

```toml
[tool.ruff.lint.isort]
required-imports = ["from __future__ import annotations"]
```

And the opposite case could potentially be covered by [unecessary-future-import](https://docs.astral.sh/ruff/rules/unnecessary-future-import/).

**Comment by Daverball** [comment]
That seems right to me, but I haven't looked at the implementation details of PEP 695 to know whether or not there are any important edge cases we would want to encode into the semantic model in some way.

Either way this seems like a good start. Although I think it would be better if we added a semantic flag to encode this property, rather than having to pass around the version, so we can have a single method on the semantic model to tell us whether or not our annotations are deferred, i.e. have a `future_or_deferred_annotations_or_stub` take the place of `future_annotations_or_stub`, since in most cases we probably want to check both.

**Comment by nickdrozd** [comment]
I am really itching to cut `from __future__ import annotations` from my code. So I was excited to try out this change with 0.11.9. But sadly it only partially works. Running into two problems.

First, some annotations are still getting eagerly evaluated:

```python
Xs: list[X] = []  # false positive F821 Undefined name `X`

x: X              # false positive F821 Undefined name `X`

x = X()           # true positive F821 Undefined name `X`

class Y:
    x: X          # false positive F821 Undefined name `X`

class X:
    x: X          # false positive F821 Undefined name `X`
```

Second, false positives for `runtime-import-in-type-checking-block (TC004)`:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Iterable  # false positive TC004

def f() -> Iterable[int]:
    return [1, 2, 3]
```

Mypy has no problem with any of this (aside from the true positive).

**Comment by nickdrozd** [comment]
Whoops! Forgot to update `target-version`. Everything seems to work. Thanks!
